### PR TITLE
Remove url-parse and punycode

### LINKED ISSUE #278: Can url-parse library be dropped?
Hello!

According to url-parse README file (https://www.npmjs.com/package/url-parse):

> url-parse was created in 2014 when the WHATWG URL API was not available in Node.js and the URL interface was supported only in some browsers. Today this is no longer true. The URL interface is available in all supported Node.js release lines and basically all browsers. Consider using it for better security and accuracy.

I wonder if url-parse is still necessary in this library, or it can be abandoned and URL be used instead. Is there any particular reason not to do this?
If this is ok, I'd be available to submit a Pull Request that does the work

### PR DESCRIPTION
**Pull request by wjhsf**
We originally added url-parse to facilitate using the library with react native. However, there's a [URL polyfill](https://github.com/charpeni/react-native-url-polyfill) that react native users can use, and we don't otherwise support the runtime, so we've decided to remove the dependency (fixes #278).

I also realized that `new URL(str)` also does the same thing as the `punycode` library, so I removed that as well.

### CODE REVIEW

File: lib/__tests__/cookieJar.spec.ts [hunk 1]
```diff
@@ -1489,25 +1496,27 @@ describe('Synchronous API on async CookieJar', () => {
...
-  it('getCookies', () => {
+  it('getCookies', (done) => {
     const invalidUrl = {}
     const cookieJar = new CookieJar()
     // @ts-expect-error deliberately trigger validation error
     void cookieJar.getCookies(invalidUrl, (err) => {
       expect(err).toMatchObject({
         message: '`url` argument is not a string or URL.',
       })
+      done()
```
Comment (colincasey) [inline lib/__tests__/cookieJar.spec.ts:1507]: good catch. of course, i fix a bug where callbacks were not being invoked and then wrote tests that fail to signal the test is done in the callback 🤦

**Review by colincasey — APPROVED** [review]

**Comment by ctomczyk** [comment]
When are these changes planned to be merged and published?

**Comment by colincasey** [comment]
@ctomczyk We can't provide you specific dates but these changes should be merged soon. Once merged we'll publish a new release candidate containing these changes for further testing. 

Please keep an eye on this PR and try out the release candidates as they become available. It can be installed as `tough-cookie@next`. We'd love to hear feedback from early adopters.

**Comment by ctomczyk** [comment]
> @ctomczyk We can't provide you specific dates but these changes should be merged soon. Once merged we'll publish a new release candidate containing these changes for further testing.
> 
> Please keep an eye on this PR and try out the release candidates as they become available. It can be installed as `tough-cookie@next`. We'd love to hear feedback from early adopters.

Understood.

Waiting for these changes in order to resolve "The `punycode` module is deprecated."

**Comment by wjhsf** [comment]
> Waiting for these changes in order to resolve "The punycode module is deprecated."

That deprecation warning is not related to this PR. See https://github.com/salesforce/tough-cookie/issues/348#issuecomment-1864824435 for details.

**Comment by ctomczyk** [comment]
> > Waiting for these changes in order to resolve "The punycode module is deprecated."
> 
> That deprecation warning is not related to this PR. See [#348 (comment)](https://github.com/salesforce/tough-cookie/issues/348#issuecomment-1864824435) for details.

Oh, I see. This is https://github.com/salesforce/tough-cookie/pull/346

**Pull request merged**